src/components/Context.js

```diff
@@ -183,49 +219,51 @@ class ProductProvider extends Component {
   addTotals = () => {
```

Comment:
delete this whole function.
the totals should be calculated inside your reducer, ideally inside that common function `calculateCart()` that I mentioned in the other comment.

The idea is to have a function that receives all information (cart items together with quantity, and a voucher) and calculates the totals/subtotals. Many different actions will rely on this function, e.g. when you add stuff to cart, when you change quantity, when you clear cart, when you apply or remove a coupon => these all should go through that single function - a central place which holds logic to calculate all sorts of totals, discounts, etc.